### PR TITLE
[agent] full reset using execvp

### PR DESCRIPTION
**Pull request by simonlingoogle**
Premature reset has causing a number of bugs.

This commit fix the bugs and avoid potential future bugs by resetting otbr-agent using `execvp`, which will bootstrap the whole process completely.

### CODE REVIEW

**Review by bukepo — APPROVED** [review]
I think  `reset` is only needed for test. I'd suggest removing this command from OpenThread and implement this function in node.py or THCI script.

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

**Comment by simonlingoogle** [comment]
@jwhui 
https://github.com/openthread/openthread/pull/6204 depends on this PR. I think we can merge this PR first and then https://github.com/openthread/openthread/pull/6204.

File: src/rest/rest_web_server.hpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ class RestWebServer
      * @retval  OTBR_ERROR_REST     Failed due to rest error .
      *
      */
-    otbrError Init(void);
+    void Init(void);
```
Comment (jwhui) [inline src/rest/rest_web_server.hpp:69]: Update docs above.

File: src/utils/socket_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#include <sys/socket.h>
+#include <unistd.h>
+
+enum SocketBlockOption
+{
+    kSocketBlock,
+    kSocketNonBlock,
+};
+
+int SocketWithCloseExec(int aDomain, int aType, int aProtocol, SocketBlockOption aBlockOption);
```
Comment (jwhui) [inline src/utils/socket_utils.hpp:47]: Add docs?

File: src/utils/socket_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
```
Comment (jwhui) [inline src/utils/socket_utils.hpp:2]:
```suggestion
 *  Copyright (c) 2021, The OpenThread Authors.
```

Comment (simonlingoogle) [reply src/utils/socket_utils.hpp:2]: It was 2020 because this PR was actually created quite a time ago.

**Comment by simonlingoogle** [comment]
> I think `reset` is only needed for test. I'd suggest removing this command from OpenThread and implement this function in node.py or THCI script.

Foremost, I think it deserves comprehensive discussion whether `reset` is a command that should exist in OpenThread, which is way beyond the scope of this PR (which just changes the behavior or reset command on OTBR).

However, there are some practical points:

1. `reset` command is required by TestHarness and many other test scripts, which can not be implemented totally in THCI or node.py. We can restart the otbr-agent process as a equivalent of `reset` but we can't do that for SOC devices. Anyway, we can't totally move the implementation of `reset` to scripts.
2. Software reset is a real thing on embedded devices. Many embedded platforms implemented `otPlatReset` for OpenThread using platform-specific **RESET** commands. So, if we can **reset** the SOC devices through OT CLI, we would also like to **reset** a otbr-agent process in the same way so that we provide consistent OT CLI user experience. 
3. If we remove `reset` from otbr-agent, it would make testing scripts (e.g. OTCI) more difficult to implement. 

Thoughts?

**Review by jwhui — APPROVED** [review]

**Pull request merged**